src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -111,26 +121,268 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
+                    ConcatIterator<TSource> previous = _previous;
+
+                    // TODO: This can be made more readable once we use C# 7 which supports pattern matching.
+                    ConcatNCollectionIterator<TSource> previousN;
+                    while ((previousN = previous as ConcatNCollectionIterator<TSource>) != null)
+                    {
+                        checked
+                        {
+                            totalCount += previousN._next.Count;
+                        }
+                        previous = previousN._previous;
+                    }
+
+                    var previous2 = previous as Concat2CollectionIterator<TSource>;
+                    Debug.Assert(previous2 != null);
+                    return checked(totalCount + previous2.Count);
+                }
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNCollectionIterator<TSource>(_previous, _next, _nextIndex);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                var nextCollection = next as ICollection<TSource>;
+                if (nextCollection != null)
+                {
+                    if (_nextIndex == int.MaxValue - 2)
+                    {
+                        // In the unlikely case of this many concatenations, if we produced a ConcatNCollectionIterator
+                        // with int.MaxValue then state would overflow before it matched it's index.
+                        // So we use the naïve approach of just having a left and right sequence.
+                        return new Concat2EnumerableIterator<TSource>(this, next);
+                    }
+
+                    return new ConcatNCollectionIterator<TSource>(this, nextCollection, _nextIndex + 1);
+                }
+                
+                // If we encounter a non-ICollection then getting .Count and performing .ToArray()
+                // will no longer be cheap, due to enumerables' lazy nature. So, fall back to using
+                // enumerable-based iterators for any further chaining.
+                return new ConcatNEnumerableIterator<TSource>(this, next, _nextIndex + 1);
+            }
+
+            // Copy all of the elements in the iterator, finishing before arrayIndex.
+            // If arrayIndex is array.Length, we'll finish copying at the end of the array.
+            // The reason we take an ending index as opposed to a starting one is because
+            // we only hold a reference to the most recently concat'd collection. So to start
+            // copying at a certain index, we'd have to re-walk the linked list of iterators
+            // all the way back to the least recent one, and repeat that for all of the
+            // collections we hold.
+            private void CopyBefore(TSource[] array, int arrayIndex)
```

Comment:
Nit: I'd suggest renaming ```arrayIndex``` to something that'll make it clear it's not the starting index, e.g. ```nextIndexAfterCopy``` or something like that.